File: go-code/refactoring_api/parameterize_function/after/charge.go
package after

import (
	"math"
)

func baseCharge(usage int) float64 {
	if usage < 0 {
		return 0
	}
	//STEP 3: Change bottomBand with the parameterized func
	return withinBand(usage, 100, 0)*.03 +
		//bottomBand(usage)*0.03 +
		//middleBand(usage)*0.05 +
		//STEP 2: Change middleBand with the parameterized func
		withinBand(usage, 200, 100)*0.05 +
		//STEP 3: Change topBand with the parameterized func
		withinBand(usage, math.MaxInt, 200)*.07
	//+ topBand(usage) * 0.07
}

func bottomBand(usage int) float64 {
	return math.Min(float64(usage), 100)
}

func middleBand(usage int) float64 {
	if usage > 100 {
		return math.Min(float64(usage), 200) - 100
	}
	return 0
}

// STEP 1: Copy and paste middleBand func and parameterize it
func withinBand(usage int, top int, bottom int) float64 {
	if usage > bottom {
		return math.Min(float64(usage), float64(top)) - float64(bottom)
	}
	return 0
}

func topBand(usage int) float64 {
	if usage > 200 {
		return float64(usage - 200)
	}
	return 0
}

File: java-code/src/main/java/refactoring/refactoring_api/parameterize_function/after/Charge.java
package refactoring.refactoring_api.parameterize_function.after;

public class Charge {
    public double baseCharge(int usage) {
      if (usage < 0) {
        return 0;
      }
      return
          //STEP 4: Use parameterized method for bottomBand
          withinBand(usage, 100, 0) +
          //bottomBand(usage) * 0.03 +

          //STEP 2: Use parameterized method for middleBand
          withinBand(usage, 200, 100) * 0.05 +
          //middleBand(usage) * 0.05 +

          //STEP 3: Use parameterized method for topBand
          withinBand(usage, Integer.MAX_VALUE, 200) * 0.07;
          //topBand(usage) * 0.07;
    }

    private int bottomBand(int usage) {
      return Math.min(usage, 100);
    }

    private int middleBand(int usage) {
      if (usage > 100) {
        return Math.min(usage, 200) - 100;
      }
      return 0;
    }

  //STEP 1: Copy and paste middleBand method and parameterize it
    private int withinBand(int usage, int top, int bottom) {
    if (usage > bottom) {
      return Math.min(usage, top) - bottom;
    }
    return 0;
  }

    private int topBand(int usage) {
      if (usage > 200) {
        return usage - 200;
      }
      return 0;
    }
  }
